### PR TITLE
etcdserver: add --max-txn-ops flag

### LINKED ISSUE #7826: Max operations in transaction
Hi, guys!

You've got a constant in server's code:

```
	// Max operations per txn list. For example, Txn.Success can have at most 128 operations,
	// and Txn.Failure can have at most 128 operations.
	MaxOpsPerTxn = 128
```

Is there any strong reason under this value? Why is it so low and why isn't it transfered from code to server's config?

### PR DESCRIPTION
**Pull request by fanminshi**
Fixes #7826

### CODE REVIEW

File: etcdmain/config.go [hunk 1]
```diff
@@ -138,6 +138,7 @@ func newConfig() *config {
 	fs.UintVar(&cfg.TickMs, "heartbeat-interval", cfg.TickMs, "Time (in milliseconds) of a heartbeat interval.")
 	fs.UintVar(&cfg.ElectionMs, "election-timeout", cfg.ElectionMs, "Time (in milliseconds) for an election to timeout.")
 	fs.Int64Var(&cfg.QuotaBackendBytes, "quota-backend-bytes", cfg.QuotaBackendBytes, "Raise alarms when backend size exceeds the given quota. 0 means use the default quota.")
+	fs.UintVar(&cfg.MaxOpsPerTxn, "max-ops-per-txn", cfg.MaxOpsPerTxn, "maximum operations per txn that etcd server allows (0 is unlimited); defaults to unlimited.")
```
Comment (heyitsanthony) [inline etcdmain/config.go]: `max-txn-ops`; the `per` is redundant. Should also probably default to 128 so it's harder to take down etcd with the default settings.

File: etcdserver/api/v3rpc/key.go [hunk 2]
```diff
@@ -39,6 +41,7 @@ type kvServer struct {
 }
 
 func NewKVServer(s *etcdserver.EtcdServer) pb.KVServer {
+	MaxOpsPerTxn = int(s.Cfg.MaxOpsPerTxn)
```
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/key.go]: this will break embedding when using multiple etcd servers with different max txn op configurations

Comment (fanminshi) [inline etcdserver/api/v3rpc/key.go]: @heyitsanthony  I am thinking about each KvServer should hold its own MaxOpsPerTxn stored in
```go
type kvServer struct {
	hdr header
	kv  etcdserver.RaftKV
        maxTxnOps unint
}
```

and use kvServer.maxTxnOps for transaction ops check?

Comment (heyitsanthony) [inline etcdserver/api/v3rpc/key.go]: ok

**Comment by fanminshi** [comment]
All fixed. PTAL

**Comment by fanminshi** [comment]
fixed CI breakage.

File: etcdserver/api/v3rpc/key.go [hunk 3]
```diff
@@ -150,8 +151,9 @@ func checkDeleteRequest(r *pb.DeleteRangeRequest) error {
 	return nil
 }
 
-func checkTxnRequest(r *pb.TxnRequest) error {
-	if len(r.Compare) > MaxOpsPerTxn || len(r.Success) > MaxOpsPerTxn || len(r.Failure) > MaxOpsPerTxn {
+func checkTxnRequest(r *pb.TxnRequest, maxTxnOps int) error {
+	plog.Infof("maxTxnOps %v", maxTxnOps)
```
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/key.go]: remove?

File: etcdmain/help.go [hunk 4]
```diff
@@ -66,6 +66,8 @@ member flags:
 		comma-separated whitelist of origins for CORS (cross-origin resource sharing).
 	--quota-backend-bytes '0'
 		raise alarms when backend size exceeds the given quota (0 defaults to low space quota).
+	--max-txn-ops '128' 
+		maximum operations per txn that etcd server allows; defaults to 128.
```
Comment (heyitsanthony) [inline etcdmain/help.go]: the default is already given in the line above, don't restate
`maximum number of operations  permitted in a transaction`

File: etcdmain/config.go [hunk 5]
```diff
@@ -138,6 +138,7 @@ func newConfig() *config {
 	fs.UintVar(&cfg.TickMs, "heartbeat-interval", cfg.TickMs, "Time (in milliseconds) of a heartbeat interval.")
 	fs.UintVar(&cfg.ElectionMs, "election-timeout", cfg.ElectionMs, "Time (in milliseconds) for an election to timeout.")
 	fs.Int64Var(&cfg.QuotaBackendBytes, "quota-backend-bytes", cfg.QuotaBackendBytes, "Raise alarms when backend size exceeds the given quota. 0 means use the default quota.")
+	fs.UintVar(&cfg.MaxTxnOps, "max-txn-ops", cfg.MaxTxnOps, "Maximum operations per txn that etcd server allows; defaults to 128.")
```
Comment (heyitsanthony) [inline etcdmain/config.go]: the default should already show up in `--help`, don't bother restating here

`Maximum number of operations permitted in a transaction.`

**Comment by fanminshi** [comment]
All fixed.

**Pull request merged**